twill-yarn/src/main/java/org/apache/twill/internal/appmaster/RunningContainers.java

```diff
@@ -113,9 +117,11 @@ public Integer apply(BitSet input) {
   private final Location applicationLocation;
   private final Set<String> runnableNames;
   private final Map<String, Map<String, String>> logLevels;
+  private final Map<String, Integer> maxRetries;
```

Comment:
but on second thought, the additional complexity of tracking by instance id would handle the following case better:

- initially start with `x` instances, all of which succeed.
- add 1 more instance, which fails

If tracking by instance id, the new instance will get maxRetries.  Without tracking by instance id, the new instance would get `maxRetries*(x+1)` retries.  So I will add the complexity of keeping track by id.  It seems worth it.